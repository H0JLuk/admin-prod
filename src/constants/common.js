export const showCount = { formatter: ({ count, maxLength }) => `Осталось символов ${maxLength - count}` };

/* '^(https?:\\/\\/)?'+ // protocol
    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
    '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
    '(\\:\\d+)?(\\/[-a-z\\d%_.~+{}]*)*'+ // port and path
    '(\\?[;&a-z\\d%_.~+=-{}]*)?'+ // query string
    '(\\#[-a-z\\d_]*)?$','i'); // fragment locator */
export const URL_REGEXP = /^(https?:\/\/)?((([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}|((\d{1,3}\.){3}\d{1,3}))(:\d+)?(\/[-A-Za-z\d%_.~+#]*)*(\?[;&a-z\d%_.~+=-{}]*)?(#[-a-z\d_]*)?$/;
