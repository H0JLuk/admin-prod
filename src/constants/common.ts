import { TextAreaProps } from 'antd/lib/input';

export const showCount: TextAreaProps['showCount'] = {
    formatter: ({ count, maxLength }) => `Осталось символов ${maxLength! - count}`,
};

export const DEFAULT_SLEEP_TIME = 500; // in milliseconds

export enum DIRECTION {
    ASC = 'ASC',
    DESC = 'DESC',
}
