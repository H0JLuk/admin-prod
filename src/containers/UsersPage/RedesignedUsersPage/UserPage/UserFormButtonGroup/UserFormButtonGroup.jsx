import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'antd';

import styles from './UserFormButtonGroup.module.css';

const BUTTON_DELETE_TEXT = 'Удалить';
const BUTTON_CANCEL_TEXT = 'Отменить';
const BUTTON_ADD_USER_TEXT = 'Добавить';
const BUTTON_SAVE_EDIT_USER_TEXT = 'Сохранить';
const INFO_USER_BUTTONS = {
    RESET_PASSWORD: 'Сбросить пароль',
    EDIT: 'Редактировать',
    UNBLOCK: 'Разблокировать',
};

const UserFormButtonGroup = ({
    type,
    onDelete,
    onCancel,
    onSubmit,
    onResetPassword,
    onEditUser,
    disableAllButtons,
    userBlocked,
}) => {
    const deleteButton = (
        <CustomButton
            className={ styles.deleteButton }
            onClick={ onDelete }
            disabled={ disableAllButtons }
        >
            { BUTTON_DELETE_TEXT }
        </CustomButton>
    );

    const cancelButton = (
        <CustomButton
            className={ styles.cancelButton }
            onClick={ onCancel }
            disabled={ disableAllButtons }
        >
            { BUTTON_CANCEL_TEXT }
        </CustomButton>
    );

    switch (type) {
        case 'new': {
            return (
                <>
                    { cancelButton }
                    <CustomButton
                        className={ styles.submitButton }
                        onClick={ onSubmit }
                        disabled={ disableAllButtons }
                    >
                        { BUTTON_ADD_USER_TEXT }
                    </CustomButton>
                </>
            );
        }
        case 'edit': {
            return (
                <>
                    { cancelButton }
                    <CustomButton
                        className={ styles.submitButton }
                        onClick={ onSubmit }
                        disabled={ disableAllButtons }
                    >
                        { BUTTON_SAVE_EDIT_USER_TEXT }
                    </CustomButton>
                    { deleteButton }
                </>
            );
        }
        case 'info': {
            return (
                <>
                    <CustomButton
                        className={ styles.submitButton }
                        onClick={ onResetPassword }
                        disabled = { disableAllButtons }
                    >
                        {userBlocked ? INFO_USER_BUTTONS.UNBLOCK : INFO_USER_BUTTONS.RESET_PASSWORD }
                    </CustomButton>
                    <CustomButton
                        className={ styles.submitButton }
                        onClick={ onEditUser }
                        disabled = { disableAllButtons }
                    >
                        { INFO_USER_BUTTONS.EDIT }
                    </CustomButton>
                    { deleteButton }
                </>
            );
        }
        default: {
            return null;
        }
    }
};

UserFormButtonGroup.propTypes = {
    type: PropTypes.oneOf(['info', 'new', 'edit']),
    onDelete: PropTypes.func,
    onCancel: PropTypes.func,
    onSubmit: PropTypes.func,
    onResetPassword: PropTypes.func,
    onEditUser: PropTypes.func,
    disableAllButtons: PropTypes.bool,
    userBlocked: PropTypes.bool,
};

UserFormButtonGroup.defaultProps = {
    type: 'info',
    onDelete: () => {},
    onCancel: () => {},
    onSubmit: () => {},
    onResetPassword: () => {},
    onEditUser: () => {},
    disableAllButtons: false,
    userBlocked: false,
};

export default UserFormButtonGroup;


const CustomButton = ({
    className,
    type,
    htmlType,
    onClick,
    disabled,
    children,
}) => (
    <Button
        className={ className }
        type={ type }
        htmlType={ htmlType }
        onClick={ onClick }
        disabled={ disabled }
    >
        { children }
    </Button>
);

CustomButton.defaultProps = {
    type: 'primary',
    htmlType: 'button',
    onClick: () => {},
};

CustomButton.propTypes = {
    className: PropTypes.string,
    type: PropTypes.string,
    shape: PropTypes.string,
    htmlType: PropTypes.string,
    size: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
};
