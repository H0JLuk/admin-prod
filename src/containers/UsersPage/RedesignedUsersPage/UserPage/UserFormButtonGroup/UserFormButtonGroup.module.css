.submitButton {
    height: 35px !important;
    padding-left: 25px !important;
    padding-right: 25px !important;
    background: #09a552 !important;
    border-color: #09a552 !important;
}

.submitButton:hover,
.submitButton:focus {
    background: #1fac6a !important;
    border-color: #1fac6a !important;
}

.cancelButton {
    height: 35px !important;
    padding-left: 25px !important;
    padding-right: 25px !important;
    background-color: white;
    color: #333f48;
    border: 1px solid #d9d9d9 !important;
}

.cancelButton:hover,
.cancelButton:focus {
    background-color: white;
    color: #333f48;
    border: 1px solid #d9d9d9 !important;
}

.deleteButton {
    height: 35px !important;
    padding-left: 25px !important;
    padding-right: 25px !important;
    background: #ec454d !important;
    border-color: #ec454d !important;
}

.deleteButton:hover,
.deleteButton:focus {
    background: #ec454d !important;
    border-color: #ec454d !important;
}
