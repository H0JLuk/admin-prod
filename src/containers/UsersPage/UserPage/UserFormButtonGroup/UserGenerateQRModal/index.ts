import UserGenerateQRModal from './UserGenerateQRModal';

export default UserGenerateQRModal;
