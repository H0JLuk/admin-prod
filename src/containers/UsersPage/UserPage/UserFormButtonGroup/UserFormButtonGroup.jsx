import { Button } from 'antd';
import React from 'react';
import PropTypes from 'prop-types';
import { LOGIN_TYPES_ENUM } from '../../../../constants/loginTypes';

export const BUTTON = {
    DELETE_TEXT: 'Удалить',
    CANCEL_TEXT: 'Отменить',
    ADD_USER_TEXT: 'Добавить',
    SAVE_EDIT_USER_TEXT: 'Сохранить',
};

export const INFO_USER_BUTTONS = {
    RESET_PASSWORD: 'Сбросить пароль',
    EDIT: 'Редактировать',
    UNBLOCK: 'Разблокировать',
};

const UserFormButtonGroup = ({
    type,
    onDelete,
    onCancel,
    onSubmit,
    onResetPassword,
    onEditUser,
    disableAllButtons,
    userBlocked,
    userLoginType,
    actionPermissions = {},
}) => {
    const { deleteUser, editUser, resetUserPassword, unlockUser } = actionPermissions;

    const deleteButton = deleteUser && (
        <Button type="primary" danger onClick={ onDelete } disabled={ disableAllButtons }>
            { BUTTON.DELETE_TEXT }
        </Button>
    );

    const cancelButton = (
        <Button type="ghost" onClick={ onCancel } disabled={ disableAllButtons }>
            { BUTTON.CANCEL_TEXT }
        </Button>
    );

    switch (type) {
        case 'new': {
            return (
                <>
                    { cancelButton }
                    <Button type="primary" onClick={ onSubmit } disabled={ disableAllButtons }>
                        { BUTTON.ADD_USER_TEXT }
                    </Button>
                </>
            );
        }
        case 'edit': {
            return (
                <>
                    { cancelButton }
                    <Button type="primary" onClick={ onSubmit } disabled={ disableAllButtons }>
                        { BUTTON.SAVE_EDIT_USER_TEXT }
                    </Button>
                    { deleteButton }
                </>
            );
        }
        case 'info': {
            const showResetOrUnlockBtn =
                userLoginType === LOGIN_TYPES_ENUM.PASSWORD &&
                (
                    (userBlocked && unlockUser) ||
                    (!userBlocked && resetUserPassword)
                );
            return (
                <>
                    { showResetOrUnlockBtn && (
                        <Button type="primary" onClick={ onResetPassword } disabled={ disableAllButtons }>
                            { userBlocked ? INFO_USER_BUTTONS.UNBLOCK : INFO_USER_BUTTONS.RESET_PASSWORD }
                        </Button>
                    ) }
                    { editUser && (
                        <Button type="primary" onClick={ onEditUser } disabled={ disableAllButtons }>
                            { INFO_USER_BUTTONS.EDIT }
                        </Button>
                    ) }
                    { deleteButton }
                </>
            );
        }
        default: {
            return null;
        }
    }
};

UserFormButtonGroup.propTypes = {
    type: PropTypes.oneOf(['info', 'new', 'edit']),
    onDelete: PropTypes.func,
    onCancel: PropTypes.func,
    onSubmit: PropTypes.func,
    onResetPassword: PropTypes.func,
    onEditUser: PropTypes.func,
    disableAllButtons: PropTypes.bool,
    userBlocked: PropTypes.bool,
};

UserFormButtonGroup.defaultProps = {
    type: 'info',
    onDelete: () => {},
    onCancel: () => {},
    onSubmit: () => {},
    onResetPassword: () => {},
    onEditUser: () => {},
    disableAllButtons: false,
    userBlocked: false,
};

export default UserFormButtonGroup;
