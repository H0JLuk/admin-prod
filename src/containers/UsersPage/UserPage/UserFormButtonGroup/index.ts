import UserFormButtonGroup from './UserFormButtonGroup';

export { BUTTON } from './UserFormButtonGroup';

export default UserFormButtonGroup;
