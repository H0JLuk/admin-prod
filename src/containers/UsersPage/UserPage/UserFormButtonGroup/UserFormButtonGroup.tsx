import React from 'react';
import { Button } from 'antd';
import noop from 'lodash/noop';
import PropTypes from 'prop-types';
import UserGenerateQRModal from './UserGenerateQRModal';
import { LOGIN_TYPES_ENUM } from '@constants/loginTypes';
import { InteractionsCurrUserToOtherUser } from '@constants/permissions';
import { ClientAppDto, UserInfo } from '@types';
import { BUTTON_TEXT } from '@constants/common';

type UserFormButtonGroupProps = {
    type?: 'info' | 'new' | 'edit' | string;
    clientApps?: ClientAppDto[];
    onDelete?: () => void;
    onCancel?: () => void;
    onSubmit?: () => void;
    onResetPassword?: () => void;
    onEditUser?: () => void;
    disableAllButtons?: boolean;
    userData?: UserInfo | null;
    actionPermissions?: InteractionsCurrUserToOtherUser;
};

export const INFO_USER_BUTTONS = {
    RESET_PASSWORD: 'Сбросить пароль',
    EDIT: 'Редактировать',
    UNBLOCK: 'Разблокировать',
    QR_CODE_GEN_TEXT: 'Сгенерировать QR-код',
};

const UserFormButtonGroup: React.FC<UserFormButtonGroupProps> = ({
    type,
    clientApps,
    onDelete,
    onCancel,
    onSubmit,
    onResetPassword,
    onEditUser,
    disableAllButtons,
    userData,
    actionPermissions,
}) => {
    const { deleteUser, editUser, resetUserPassword, unlockUser, canGenerateQRCode } = actionPermissions || {};
    const { tmpBlocked, personalNumber, loginType, clientAppIds, id } = userData || {};

    const deleteButton = deleteUser && (
        <Button
            type="primary"
            onClick={onDelete}
            disabled={disableAllButtons}
            danger
        >
            {BUTTON_TEXT.DELETE}
        </Button>
    );

    const cancelButton = (
        <Button
            type="ghost"
            onClick={onCancel}
            disabled={disableAllButtons}
        >
            {BUTTON_TEXT.CANCEL}
        </Button>
    );

    switch (type) {
        case 'new': {
            return (
                <>
                    {cancelButton}
                    <Button
                        type="primary"
                        onClick={onSubmit}
                        disabled={disableAllButtons}
                    >
                        {BUTTON_TEXT.ADD}
                    </Button>
                </>
            );
        }
        case 'edit': {
            return (
                <>
                    {cancelButton}
                    <Button
                        type="primary"
                        onClick={onSubmit}
                        disabled={disableAllButtons}
                    >
                        {BUTTON_TEXT.SAVE}
                    </Button>
                    {deleteButton}
                </>
            );
        }
        case 'info': {
            const showResetOrUnlockBtn =
                loginType === LOGIN_TYPES_ENUM.PASSWORD &&
                (
                    (tmpBlocked && unlockUser) ||
                    (!tmpBlocked && resetUserPassword)
                );
            const showQRCodeBtn = loginType === LOGIN_TYPES_ENUM.DIRECT_LINK && canGenerateQRCode;
            return (
                <>
                    {showQRCodeBtn && (
                        <UserGenerateQRModal
                            clientApps={clientApps}
                            personalNumber={personalNumber}
                            userId={id}
                            userClientAppIds={clientAppIds}
                            buttonDisabled={disableAllButtons}
                        />
                    )}
                    {showResetOrUnlockBtn && (
                        <Button
                            type="primary"
                            onClick={onResetPassword}
                            disabled={disableAllButtons}
                        >
                            {tmpBlocked ? INFO_USER_BUTTONS.UNBLOCK : INFO_USER_BUTTONS.RESET_PASSWORD}
                        </Button>
                    )}
                    {editUser && (
                        <Button
                            type="primary"
                            onClick={onEditUser}
                            disabled={disableAllButtons}
                        >
                            {INFO_USER_BUTTONS.EDIT}
                        </Button>
                    )}
                    {deleteButton}
                </>
            );
        }
        default: {
            return null;
        }
    }
};

UserFormButtonGroup.propTypes = {
    type: PropTypes.oneOf(['info', 'new', 'edit']),
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    clientApps: PropTypes.array,
    onDelete: PropTypes.func,
    onCancel: PropTypes.func,
    onSubmit: PropTypes.func,
    onResetPassword: PropTypes.func,
    onEditUser: PropTypes.func,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    userData: PropTypes.object,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    actionPermissions: PropTypes.object,
    disableAllButtons: PropTypes.bool,
};

UserFormButtonGroup.defaultProps = {
    type: 'info',
    onDelete: noop,
    onCancel: noop,
    onSubmit: noop,
    onResetPassword: noop,
    onEditUser: noop,
    disableAllButtons: false,
};

export default UserFormButtonGroup;
