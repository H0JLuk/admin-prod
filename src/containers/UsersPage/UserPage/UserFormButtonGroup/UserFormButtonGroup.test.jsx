import React from 'react';
import { mount, shallow } from 'enzyme';
import UserFormButtonGroup, { INFO_USER_BUTTONS } from './UserFormButtonGroup';
import { LOGIN_TYPES_ENUM } from '../../../../constants/loginTypes';
import { userTestData } from '../../../../../__tests__/constants';
import { BUTTON_TEXT } from '@constants/common';

const CASE_TYPES = {
    NEW: 'new',
    EDIT: 'edit',
    INFO: 'info',
    NONE: '',
};

const LOGIN_TYPE = LOGIN_TYPES_ENUM.PASSWORD;

const TEST_ACTION_PROPS = {
    deleteUser: true,
    editUser: true,
    resetUserPassword: true,
    unlockUser: true,
    canGenerateQRCode: true,
};

const TEST_ACTION_PROPS_FALSE = {
    deleteUser: false,
    editUser: false,
    resetUserPassword: false,
    unlockUser: false,
    canGenerateQRCode: false,
};

const TEST_PROPS = {
    blocked: false,
    disableBtn: false,
    onDelete: jest.fn(),
    onCancel: jest.fn(),
    onSubmit: jest.fn(),
    onResetPassword: jest.fn(),
    onEditUser: jest.fn(),
    userData: userTestData,
    userLoginType: LOGIN_TYPE,
    actionPermissions: TEST_ACTION_PROPS,
};

describe('<UserFormButtonGroup /> test', () => {
    const Component = (type = CASE_TYPES.EDIT, tmpBlocked = false) => mount(
        <UserFormButtonGroup
            { ...TEST_PROPS }
            userData={ { ...TEST_PROPS.userData, tmpBlocked } }
            type={ type }
        />
    );
    const ComponentEdit = Component();
    const ComponentButtons = ComponentEdit.find('button');

    it('should be mount snap', () => {
        expect(ComponentEdit.html()).toMatchSnapshot();
    });

    it('should be visible type edit buttons and call functions', () => {
        expect(ComponentButtons.at(0).text()).toBe(BUTTON_TEXT.CANCEL);
        ComponentButtons.at(0).simulate('click');
        expect(TEST_PROPS.onCancel).toBeCalledTimes(1);

        expect(ComponentButtons.at(1).text()).toBe(BUTTON_TEXT.SAVE);
        ComponentButtons.at(1).simulate('click');
        expect(TEST_PROPS.onSubmit).toBeCalledTimes(1);

        expect(ComponentButtons.at(2).text()).toBe(BUTTON_TEXT.DELETE);
        ComponentButtons.at(2).simulate('click');
        expect(TEST_PROPS.onDelete).toBeCalledTimes(1);
    });

    it('should be visible type new buttons and call functions', () => {
        const ComponentNew = Component(CASE_TYPES.NEW);
        const ComponentNewButtons = ComponentNew.find('button');

        expect(ComponentNewButtons.at(0).text()).toBe(BUTTON_TEXT.CANCEL);
        ComponentNewButtons.at(0).simulate('click');
        expect(TEST_PROPS.onCancel).toBeCalledTimes(1);

        expect(ComponentNewButtons.at(1).text()).toBe(BUTTON_TEXT.ADD);
        ComponentNewButtons.at(1).simulate('click');
        expect(TEST_PROPS.onSubmit).toBeCalledTimes(1);
    });

    it('should be visible type info buttons and call functions with edit user', () => {
        const ComponentInfo = Component(CASE_TYPES.INFO);
        const ComponentInfoButtons = ComponentInfo.find('button');

        expect(ComponentInfoButtons.at(0).text()).toBe(INFO_USER_BUTTONS.RESET_PASSWORD);
        ComponentInfoButtons.at(0).simulate('click');
        expect(TEST_PROPS.onResetPassword).toBeCalledTimes(1);

        expect(ComponentInfoButtons.at(1).text()).toBe(INFO_USER_BUTTONS.EDIT);
        ComponentInfoButtons.at(1).simulate('click');
        expect(TEST_PROPS.onEditUser).toBeCalledTimes(1);

        expect(ComponentInfoButtons.at(2).text()).toBe(BUTTON_TEXT.DELETE);
        ComponentInfoButtons.at(2).simulate('click');
        expect(TEST_PROPS.onDelete).toBeCalledTimes(1);
    });

    it('should be visible button text unblock user', () => {
        const ComponentInfo = Component(CASE_TYPES.INFO, true);
        const ComponentInfoButtons = ComponentInfo.find('button');
        expect(ComponentInfoButtons.at(0).text()).toBe(INFO_USER_BUTTONS.UNBLOCK);
    });

    it('should render generateQR button', () => {
        const wrapper = shallow(
            <UserFormButtonGroup
                { ...TEST_PROPS }
                userData={ {
                    ...TEST_PROPS.userData,
                    loginType: LOGIN_TYPES_ENUM.DIRECT_LINK,
                } }

                type="info"
            />
        );

        expect(wrapper.find('UserGenerateQRModal')).toHaveLength(1);
    });

    it('should be visible type info with action permissions false and 0 buttons', () => {
        const ComponentInfo = Component().setProps({ type: CASE_TYPES.INFO, actionPermissions: TEST_ACTION_PROPS_FALSE });
        const ComponentInfoButtons = ComponentInfo.find('button');

        expect(ComponentInfoButtons).toHaveLength(0);
    });

    it('should be null if no matching type', () => {
        jest.spyOn(console, 'error').mockImplementation(() => '');
        const ComponentNull = Component(CASE_TYPES.NONE);
        expect(ComponentNull.html()).toBeNull();
        expect(console.error).toBeCalled();
    });

});
