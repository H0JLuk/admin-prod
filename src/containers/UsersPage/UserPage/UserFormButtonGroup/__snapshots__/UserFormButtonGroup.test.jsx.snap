// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserFormButtonGroup /> test should be mount snap 1`] = `"<button type=\\"button\\" class=\\"ant-btn ant-btn-ghost\\"><span>Отменить</span></button><button type=\\"button\\" class=\\"ant-btn ant-btn-primary\\"><span>Сохранить</span></button><button type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-dangerous\\"><span>Удалить</span></button>"`;
